src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1558,7 +1609,17 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 
 	foreach_oid(relationId, options->relationIdList)
 	{
-		AcquireColocationLock(relationId, operationName);
+		AcquireColocationLock(relationId, ExclusiveLock, operationName);
+
+		/*
+		 * We release the lock here, and the lower level UDF will acquire it back.
+		 * If in the meantime another process acquires the lock, the lower level
+		 * UDF will bail out anyway.
+		 *
+		 * We have to release the lock because the lower level UDF is executed via
+		 * another backend (e.g., otherwise this process conflicts with it).
```

Comment:
Otherwise it is conflicting with the same lock on `citus_move_shard_placement` running via another session.

Reply:
but why do we take the lock?

Reply:
btw, I'm not really happy with this part of the code. Now, experimenting to use maybe another lock. Let me see how that'd look

Reply:
in any case, taking a lock and then immediately releasing it does not seem helpful

Reply:
I think the main purpose is to error early and clearly if another rebalance is running.

Reply:
> I think the main purpose is to error early and clearly if another rebalance is running.

yeh, but we could easily have 2 rebalance operations pass this point at the same time if the lock is released immediately. We could also have a rebalance starting in the middle of another rebalance (in between moves).

It seems like we should have a rebalance lock.

Reply:
The problem with removing is that:

Session 1: only rebalances colocation group 2
```SQL
SELECT rebalance_table_shards('colocation_2');
```

-- session 2: rebalances all colocations
```SQL
SELECT rebalance_table_shards();
```

Would not throw an error as it used to be. There are lots of tests similar to this which fails: https://github.com/citusdata/citus/blob/main/src/test/regress/spec/isolation_shard_rebalancer.spec


Maybe they are not very critical, but not having the lock would mean we'd break the user experience, and we need to backport this.

Reply:
In that case they will still fail once they call the shard move step, right?

Reply:
But I agree there would still be theoretical race conditions then. So, it would be good to have two locks:
1. A rebalance lock
2. A shard move lock

Reply:
> In that case they will still error once they call the shard move step, right?

Not really, because before acquiring the same colocationId lock, the 2. rebalance blocks on pg_dist_node. So, we do not throw an error until the first one commits, which also doesn't seem to fit the `fail early` design

Reply:
Ok, so I pushed my latest change (missing comment / better naming / the test on the #5060).

I think this is the least intrusive change we can backport:
- create dist table acquires `ShareLock` on `ColocationPlacementLock`
- `citus_move_shard_placement` acquires `ExclusiveLock` on `ColocationPlacementLock`
- `rebalance_table_shard` acquires `ExclusiveLock` on `ColocationRelationLock`


Basically,
-  prevents any kind of concurrent shard move/copy operation
- prevents concurrent table creation with any kind of concurrent shard move/copy operation
- allows concurrent create_dist_table -- not fulfills @marcocitus's comment. But, maybe that is for another PR because it seems hard to backport. Also, triggers lots of regression test failures.